# Jumpman23 - Data Cleaning

Jumpman23 is an on-demand food delivery platform connecting customers to “Jumpman”, a vast network of couriers. Jumpman23 recently launched in its newest market, New York City.

In this Jupyter Notebook, we will focus on cleaning the data so it's ready for analysis. The cleaned data will be exported into a new csv file, which will be used in the next Jupter Notebook for analysis. 

In [1]:
import pandas as pd
import numpy as np
import datetime as dt

In [2]:
#read the csv file
df = pd.read_csv("analyze_me.csv")
df = df.sort_values('when_the_delivery_started') #sort by chronoligical order
df

delivery_id  customer_id  jumpman_id vehicle_type  \
2087      1271706        66556      154943          car   
4946      1271751       104131      124897      bicycle   
5719      1271867       121706      124897      bicycle   
2296      1272279       304449      112830      bicycle   
4009      1272303       107621      127812      bicycle   
...           ...          ...         ...          ...   
3962      1491147       320359      172130       walker   
5101      1491144        54851      104533      bicycle   
3488      1491341       229946       59348       walker   
1988      1491424       391367      172130       walker   
1315      1491424       391367      172130       walker   

                  pickup_place     place_category  \
2087              CVS Pharmacy                NaN   
4946               Duane Reade         Drug Store   
5719                McDonald's          Fast Food   
2296        Friend of a Farmer           American   
4009     Duane Reade duplicate  Convenience Store   
...                        ...                ...   
3962  Taqueria Lower East Side                NaN   
5101            RedFarm Hudson            Chinese   
3488         The Meatball Shop            Italian   
1988                   Veselka            Russian   
1315                   Veselka            Russian   

                           item_name  item_quantity  \
2087                             NaN            NaN   
4946                             NaN            NaN   
5719              World Famous Fries            2.0   
2296                           Bacon            1.0   
4009                             NaN            NaN   
...                              ...            ...   
3962                             NaN            NaN   
5101  ‘Pac Man’ Shrimp Dumplings (4)            1.0   
3488                Sautéed Broccoli            1.0   
1988         Small Plate of Pierogis            1.0   
1315         Small Plate of Pierogis            1.0   

               item_category_name how_long_it_took_to_order  pickup_lat  \
2087                          NaN                       NaN   40.744157   
4946                          NaN                       NaN   40.763217   
5719               Snacks & Sides                       NaN   40.756799   
2296                        Sides            00:09:08.75807   40.736794   
4009                          NaN                       NaN   40.743171   
...                           ...                       ...         ...   
3962                          NaN           00:02:51.502102   40.722341   
5101                      Dim Sum           00:04:54.552157   40.734214   
3488                 Green Market           00:03:42.304071   40.721545   
1988  Handmade Varenyky (Pierogi)           00:07:29.766955   40.729136   
1315  Handmade Varenyky (Pierogi)           00:07:29.766955   40.729136   

      pickup_lon  dropoff_lat  dropoff_lon   when_the_delivery_started  \
2087  -73.991987    40.744079   -73.991519  2014-10-01 00:07:58.632482   
4946  -73.967613    40.765340   -73.967106  2014-10-01 00:26:31.924774   
5719  -73.987899    40.756804   -73.998153   2014-10-01 01:00:06.75635   
2296  -73.986800    40.733020   -73.982436  2014-10-01 08:46:15.935061   
4009  -73.984004    40.744964   -73.985617  2014-10-01 09:20:21.573801   
...          ...          ...          ...                         ...   
3962  -73.987829    40.724376   -73.994179  2014-10-30 22:31:58.003417   
5101  -74.006202    40.744761   -74.003630  2014-10-30 22:32:24.293206   
3488  -73.988842    40.724556   -73.991582   2014-10-30 22:56:00.07339   
1988  -73.986972    40.723531   -73.982324    2014-10-30 23:08:43.4819   
1315  -73.986972    40.723531   -73.982324    2014-10-30 23:08:43.4819   

     when_the_Jumpman_arrived_at_pickup when_the_Jumpman_left_pickup  \
2087                                NaN                          NaN   
4946         2014-10-01 00:39:31.086322   2014-10-01 00:59:57.522

There are 5983 rows and 18 columns in this dataset. The dates range between 2014-10-01 to 2014-10-30. The data set contains information related to the time and location of the pickup and dropoff, transportation mode used by the Jumpman, as well as categorical data related to the merchant.

Let's begin cleaning the data!

In [3]:
df.dtypes

delivery_id                              int64
customer_id                              int64
jumpman_id                               int64
vehicle_type                            object
pickup_place                            object
place_category                          object
item_name                               object
item_quantity                          float64
item_category_name                      object
how_long_it_took_to_order               object
pickup_lat                             float64
pickup_lon                             float64
dropoff_lat                            float64
dropoff_lon                            float64
when_the_delivery_started               object
when_the_Jumpman_arrived_at_pickup      object
when_the_Jumpman_left_pickup            object
when_the_Jumpman_arrived_at_dropoff     object
dtype: object

### Formatting the Data

Using the time data, we can calculate the duration at each step of the order. 

 - how_long_it_took_to_order_mins - length of time the customer took to place the order
 - wait_at_rest - length of the time between when the jumpman arrived at pickup and left the pickup
 - prep_time - duration between when the order was placed by the customer and when the jumpman left the pickup
 - transit_time - duration between when the jumpman picked up the order and dropped it off to the customer
 - total_time - the total duration of the order from start to finish

First, we'll need to convert the time columns to datatime objects. The column, how_long_it_took_to_order, is in a timestamp format, which is not very useful. By converting it to a timedelta, we can extract the time in minutes.


In [4]:
#convert time to datetime obj
df['when_the_Jumpman_arrived_at_dropoff'] = pd.to_datetime(df['when_the_Jumpman_arrived_at_dropoff'])
df['when_the_Jumpman_left_pickup'] = pd.to_datetime(df['when_the_Jumpman_left_pickup'])
df['when_the_Jumpman_arrived_at_pickup'] = pd.to_datetime(df['when_the_Jumpman_arrived_at_pickup'])
df['when_the_delivery_started'] = pd.to_datetime(df['when_the_delivery_started'])
# convert timestamp to timedelta, then extract minutes
df['how_long_it_took_to_order_mins'] = pd.to_timedelta(df['how_long_it_took_to_order'], errors='coerce').apply(lambda x :(x.total_seconds()-x.days*86400))/60

#Calculate durations
#TOTAL wait time
df['total_time'] = df['when_the_Jumpman_arrived_at_dropoff'] - df['when_the_delivery_started']
df['total_time'] = df['total_time'].apply(lambda x: x.total_seconds()/60)

#TRANSIT time
df['transit_time'] = df['when_the_Jumpman_arrived_at_dropoff'] - df['when_the_Jumpman_left_pickup']
df['transit_time'] = df['transit_time'].apply(lambda x: x.total_seconds()/60)

#PREP time
df['prep_time'] = df['when_the_Jumpman_left_pickup'] - df['when_the_delivery_started']
df['prep_time'] = df['prep_time'].apply(lambda x: x.total_seconds()/60)

#Wait time at pickup
df['wait_at_rest'] = df['when_the_Jumpman_left_pickup'] - df['when_the_Jumpman_arrived_at_pickup']
df['wait_at_rest'] = df['wait_at_rest'].apply(lambda x: x.total_seconds()/60)


df.head()

delivery_id  customer_id  jumpman_id vehicle_type  \
2087      1271706        66556      154943          car   
4946      1271751       104131      124897      bicycle   
5719      1271867       121706      124897      bicycle   
2296      1272279       304449      112830      bicycle   
4009      1272303       107621      127812      bicycle   

               pickup_place     place_category           item_name  \
2087           CVS Pharmacy                NaN                 NaN   
4946            Duane Reade         Drug Store                 NaN   
5719             McDonald's          Fast Food  World Famous Fries   
2296     Friend of a Farmer           American               Bacon   
4009  Duane Reade duplicate  Convenience Store                 NaN   

      item_quantity item_category_name how_long_it_took_to_order  ...  \
2087            NaN                NaN                       NaN  ...   
4946            NaN                NaN                       NaN  ...   
5719            2.0     Snacks & Sides                       NaN  ...   
2296            1.0              Sides            00:09:08.75807  ...   
4009            NaN                NaN                       NaN  ...   

      dropoff_lon  when_the_delivery_started  \
2087   -73.991519 2014-10-01 00:07:58.632482   
4946   -73.967106 2014-10-01 00:26:31.924774   
5719   -73.998153 2014-10-01 01:00:06.756350   
2296   -73.982436 2014-10-01 08:46:15.935061   
4009   -73.985617 2014-10-01 09:20:21.573801   

      when_the_Jumpman_arrived_at_pickup  when_the_Jumpman_left_pickup  \
2087                                 NaT                           NaT   
4946          2014-10-01 00:39:31.086322    2014-10-01 00:59:57.522402   
5719          2014-10-01 01:19:29.205722    2014-10-01 01:36:49.131316   
2296          2014-10-01 09:02:40.003541    2014-10-01 09:15:59.607582   
4009          2014-10-01 09:26:01.194532    2014-10-01 09:37:56.158669   

     when_the_Jumpman_arrived_at_dropoff how_long_it_took_to_order_mins  \
2087          2014-10-01 00:30:21.109149                            NaN   
4946          2014-10-01 01:04:14.355157                            NaN   
5719          2014-10-01 01:49:29.034932                            NaN   
2296          2014-10-01 09:28:40.095456                       9.145968   
4009          2014-10-01 09:39:41.631246                            NaN   

     total_time transit_time  prep_time  wait_at_rest  
2087  22.374611          NaN        NaN           NaN  
4946  37.707173     4.280546  33.426627     20.440601  
5719  49.371310    12.665060  36.706249     17.332093  
2296  42.402673    12.674798  29.727875     13.326734  
4009  19.334291     1.757876  17.576414     11.916069  

[5 rows x 23 columns]

### Dropping Duplicate Rows

To clean the data, we need to remove any duplicate rows. Additionally, we expect delivery_id to be unique for each order, so we'll need to check for duplicates, and remove them if there are any.

In [5]:
#check for number of duplicate rows
print(len((df[df.duplicated(keep=False)])))

# Drop any duplicate rows from df_clean
df_clean = df.copy()
df_clean.drop_duplicates(keep = 'first', inplace = True) 


32


In [6]:
#look for places where delivery_id is duplicated
(df_clean[df_clean.duplicated(subset=['delivery_id'], keep=False)]).sort_values('delivery_id')


delivery_id  customer_id  jumpman_id vehicle_type         pickup_place  \
5080      1272701        81085      112646      bicycle   Mighty Quinn's BBQ   
1008      1272701        81085      112646      bicycle   Mighty Quinn's BBQ   
2272      1274248       208020       60149          car     Murray's Falafel   
2299      1274248       208020       60149          car     Murray's Falafel   
2986      1274248       208020       60149          car     Murray's Falafel   
...           ...          ...         ...          ...                  ...   
3614      1490188       166368      174143   motorcycle  Prosperity Dumpling   
782       1490188       166368      174143   motorcycle  Prosperity Dumpling   
4074      1490744        52256       38597      bicycle          Han Dynasty   
4983      1490744        52256       38597      bicycle          Han Dynasty   
3020      1490744        52256       38597      bicycle          Han Dynasty   

      place_category                                       item_name  \
5080             BBQ                              Housemade Iced Tea   
1008             BBQ                                         Brisket   
2272  Middle Eastern  Blue Lamoon Citrus blossom lemonade w/ Splenda   
2299  Middle Eastern                          Moroccan Cigars (5 pc)   
2986  Middle Eastern                                      Watermelon   
...              ...                                             ...   
3614         Chinese               Chives and Pork Dumplings in Soup   
782          Chinese                                  Sesame Pancake   
4074         Chinese                   Bok Choy with Black Mushrooms   
4983         Chinese                                  Dan Dan Noodle   
3020         Chinese                         Wonton in Chicken Broth   

      item_quantity item_category_name how_long_it_took_to_order  ...  \
5080            1.0          Beverages                       NaN  ...   
1008            1.0              Meats                       NaN  ...   
2272            1.0          Beverages           00:07:08.767432  ...   
2299            1.0         Appetizers           00:07:08.767432  ...   
2986            1.0           Desserts           00:07:08.767432  ...   
...             ...                ...                       ...  ...   
3614            1.0          Dumplings                       NaN  ...   
782             1.0      Noodle & Buns                       NaN  ...   
4074            1.0         Vegetables           00:09:51.159698  ...   
4983            1.0            Noodles           00:09:51.159698  ...   
3020            1.0              Soups           00:09:51.159698  ...   

      dropoff_lon  when_the_delivery_started  \
5080   -73.993393 2014-10-01 12:12:24.393054   
1008   -73.993393 2014-10-01 12:12:24.393054   
2272   -73.990922 2014-10-01 17:25:48.546330   
2299   -73.990922 2014-10-01 17:25:48.546330   
2986   -73.990922 2014-10-01 17:25:48.546330   
...           ...                        ...   
3614   -73.947845 2014-10-30 21:00:30.703848   
782    -73.947845 2014-10-30 21:00:30.703848   
4074   -73.987752 2014-10-30 21:44:05.205404   
4983   -73.987752 2014-10-30 21:44:05.205404   
3020   -73.987752 2014-10-30 21:44:05.205404   

      when_the_Jumpman_arrived_at_pickup  when_the_Jumpman_left_pickup  \
5080          2014-10-01 12:17:22.929789    2014-10-01 12:27:42.369732   
1008          2014-10-01 12:17:22.929789    2014-10-01 12:27:42.369732   
2272          2014-10-01 17:40:32.886964    2014-10-01 17:53:54.166799   
2299          2014-10-01 17:40:32.886964    2014-10-01 17:53:54.166799   
2986          2014-10-01 17:40:32.886964    2014-10-01 17:53:54.166799   
...                                  ...                           ...   
3614          2014-10-30 21:08:00.931773    2014-10-30 21:44:21.145721   
782           2014-10-30 21:08:00.931773    2014-10-30 21:44:21.145721   
4074          2014-10-30 21:51:58.394867    2014-10-30 22:06

There are 1398 rows with duplicate delivery_id values! Taking a look at each order more closely, we can see that each delivery is split into multiple rows when the customer orders more than one item. Although it's interesting to note all the items customers have ordered, keeping these rows will affect our analysis later on when we calculate the total number of orders. We will keep rows with the first delivery_id occurence, and drop the rest. 

In [9]:
#keep rows having the first delivery id occurence
df_clean.drop_duplicates(subset = ['delivery_id'], keep = 'first', inplace = True) 


### Time Violations

The last type of cleaning we will do with the data is check for any time violations. For example, we can check to see if there are any deliveries where the dropoff time occured before the pickup time, which wouldn't make any sense. We will NOT be removing these rows because we will assume that the information from the other are still accurate, and can be used in our analysis. 

In [12]:
#Check for time violations

#commented out conditions that have 0 violations
#df_clean[df_clean['when_the_Jumpman_arrived_at_dropoff'] < df_clean['when_the_Jumpman_left_pickup']]
#df_clean[df_clean['when_the_Jumpman_arrived_at_dropoff'] < df_clean['when_the_Jumpman_arrived_at_pickup']]
#df_clean[df_clean['when_the_Jumpman_arrived_at_dropoff'] < df_clean['when_the_delivery_started']]
#df_clean[df_clean['when_the_Jumpman_left_pickup'] < df_clean['when_the_Jumpman_arrived_at_pickup']]


print(len(df_clean[df_clean['when_the_Jumpman_arrived_at_pickup'] < df_clean['when_the_delivery_started']]))
print(len(df_clean[df_clean['when_the_Jumpman_left_pickup'] < df_clean['when_the_delivery_started']]))


438
2


There are two time violations that occured: the jumpman appeared at the pickup before the delivery started 438 times, and the jumpman left the pickup before the delivery started two times. This may suggest that Jumpman23 has a bug in their system which incorrectly records the start time.

### Null Values

Lastly, we'll check for any columns with Null values. The most important columns that cannot have Null values are delivery_id, customer_id, and jumpman_id, and we see below that they all do not, so we won't be dropping any rows. 

In [13]:
df_clean.isnull().sum()

delivery_id                               0
customer_id                               0
jumpman_id                                0
vehicle_type                              0
pickup_place                              0
place_category                          841
item_name                              1230
item_quantity                          1230
item_category_name                     1230
how_long_it_took_to_order              2635
pickup_lat                                0
pickup_lon                                0
dropoff_lat                               0
dropoff_lon                               0
when_the_delivery_started                 0
when_the_Jumpman_arrived_at_pickup      495
when_the_Jumpman_left_pickup            495
when_the_Jumpman_arrived_at_dropoff       0
how_long_it_took_to_order_mins         2635
total_time                                0
transit_time                            495
prep_time                               495
wait_at_rest                    

Our data is finally clean and ready for analysis! We will output the cleaned data into a new csv and continue the analysis in the next Jupyter Notebook.

In [16]:
#write to csv
df_clean.to_csv('cleaned_data.csv')